dewy/common/collection_embeddings.py

```diff
@@ -0,0 +1,259 @@
+
+        Returns:
+        List of `(chunk_id, score)` pairs from the embeddings.
+        """
+        embedded_query = await self._embedding.aget_text_embedding(query)
+
+        async with self._pg_pool.acquire() as conn:
+            logger.info("Executing SQL query for chunks from {}", self.collection_id)
+            embeddings = await conn.fetch(self._retrieve_embeddings, embedded_query, n)
+            embeddings = [e["chunk_id"] for e in embeddings]
+            return embeddings
+
+    async def retrieve_text_chunks(self, query: str, n: int = 10) -> List[TextChunk]:
+        """Retrieve embeddings related to the given query.
+
+        Parameters:
+        - query: The query to retrieve matching embeddings for.
+        - n: The number of embeddings to retrieve.
+
+        Returns:
+        List of chunk_ids from the embeddings.
+        """
+        embedded_query = await self._embedding.aget_text_embedding(query)
+
+        async with self._pg_pool.acquire() as conn:
+            logger.info("Executing SQL query for chunks from {}", self.collection_id)
+            embeddings = await conn.fetch(self._retrieve_chunks, embedded_query, n)
+            embeddings = [
+                TextChunk(raw=True, score=e["score"], text=e["text"])
+                for e in embeddings
+            ]
+            return embeddings
+
+    async def ingest(self, document_id: int, url: str) -> None:
+        logger.info("Loading content for document {} from '{}'", document_id, url)
+        extracted = await extract(
+            url, extract_tables=self.extract_tables, extract_images=self.extract_images
+        )
+        if extracted.is_empty():
+            logger.error(
+                "No content retrieved from for document {} from '{}'", document_id, url
+            )
+            return
+
+        logger.info(
+            "Chunking text of length {} for {}", len(extracted.text), document_id
+        )
+
+        # Extract chunks (snippets) and perform the direct embedding.
+        text_chunks = await self._chunk_sentences(extracted.text)
+
+        logger.info("Chunking produced {} chunks for {}", len(text_chunks), document_id)
+
+        # TODO: support non-text chunks
+        # TODO: support non-snippet text chunks (eg., summary values)
+        # TODO: support indirect embeddings
+        async with self._pg_pool.acquire() as conn:
+            async with conn.transaction():
+                # First, insert the chunks.
+                await conn.executemany(
+                    """
+                    INSERT INTO chunk (document_id, kind, text)
+                    VALUES ($1, $2, $3);
+                    """,
+                    [(document_id, "text", text_chunk) for text_chunk in text_chunks],
+                )
+
+                # Then, embed each of those chunks.
+                # We assume no chunks for the document existed before, so we can iterate
+                # over the chunks.
```

Comment:
Can we not return the chunk ID's or something? This seems like an assumption that's going to cause bugs as soon as we support updating a document.

Reply:
We could. My thinking was that we could write them into the DB rather than trying to keep them in memory and then read them back out. But, I think that both llamaindex and various other embeddings will lead to the whole text having to fit in memory anyway during an ingest, so maybe it doesn't matter.

Reply:
Actually, I take that back. There isn't a great way to do that. Specifically:

1. This uses executemany, which *doesn't* return anything.
2. If we use `fetch`, we can't provide a list of rows to insert -- it needs to be a single query.

I think I'll leave as is for this PR. I think we could handle update in a variety of ways:

1. Introduce a new document ID and delete the old one.
2. Add a "version" to each chunk, and query for only the chunks related to the current version.
3. etc.

Reply:
yeah, including an "ingest version" or something that we could filter on the other side would work.